internal/slave/service.go

```diff
@@ -197,3 +238,138 @@ func newErrorStatus(err error) *serverpb.Status {
+	if ss.replLag > ss.replConfig.MaxActiveReplLag {
+		ss.regionInfo.Status = serverpb.RegionStatus_INACTIVE
+	} else if ss.lastReplTime == 0 || hlc.GetTimeAgo(ss.lastReplTime) > ss.replConfig.MaxActiveReplElapsed {
+		ss.regionInfo.Status = serverpb.RegionStatus_INACTIVE
+	} else if ss.isClosed {
+		ss.regionInfo.Status = serverpb.RegionStatus_INACTIVE
+	} else {
+		ss.regionInfo.Status = serverpb.RegionStatus_ACTIVE_SLAVE
+	}
+	ss.regionInfo.MasterHost = &ss.replConfig.ReplMasterAddr
+	ss.lg.Debug("Current Info", zap.String("Status", ss.regionInfo.Status.String()),
+		zap.Uint64("Repl Lag", ss.replLag), zap.Uint64("Last Repl time", ss.lastReplTime))
+	return ss.regionInfo, nil
+}
+
+func (ss *slaveService) replaceMasterIfInactive() error {
+	if ss.replConfig.DisableAutoMasterDisc {
+		return nil
+	}
+	if regions, err := ss.clusterInfo.GetClusterStatus(ss.regionInfo.GetDatabase(), ss.regionInfo.GetVBucket()); err == nil {
+		currentMasterIdx := len(regions)
+		for i, region := range regions {
+			if region.NodeAddress == ss.replConfig.ReplMasterAddr {
+				currentMasterIdx = i
+				break
+			}
+		}
+
+		if currentMasterIdx == len(regions) {
+			// Current Master not found in cluster. implies current master is inactive
+			return ss.reconnectMaster()
+		} else if !isVBucketApplicableToBeMaster(regions[currentMasterIdx]) {
+			return ss.reconnectMaster()
+		} else {
+			// current master is active. No action required as replication error could be temporary
+			// need to validate this assumption though
+			return nil
+		}
+	} else {
+		return err
+	}
+}
+
+func (ss *slaveService) reconnectMaster() error {
+	// Close any existing client connection
+	// This is so that replication doesn't happen from inactive master
+	// which could otherwise result in slave marking itself active if no errors in replication
+	if ss.replCli != nil {
+		ss.replCli.Close()
+		ss.replCli = nil
+		ss.replConfig.ReplMasterAddr = ""
+	}
+	return ss.findAndConnectToMaster()
+}
+
+func (ss *slaveService) findAndConnectToMaster() error {
+	if master, err := ss.findNewMaster(); err == nil {
+		// TODO: Check if authority override option is needed for slaves while they connect with masters
+		if replCli, err := ctl.NewInSecureDKVClient(master, ""); err == nil {
+			// concurrency issues ?
+			ss.replCli = replCli
+			ss.replConfig.ReplMasterAddr = master
+		} else {
+			ss.lg.Warn("Unable to create a replication client", zap.Error(err))
+			return err
+		}
+	} else {
+		ss.lg.Warn("Unable to find a master for this slave to replicate from", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+// Finds a new active master for the region
+// Prefers followers within the local DC first, followed by master within local DC
+// followed by followers outside DC, followed by master outside DC
+// TODO - rather than randomly selecting a master from applicable followers, load balance to distribute better
+func (ss *slaveService) findNewMaster() (string, error) {
```

Comment:
instead of returning a string, maybe return the object?

Reply:
you mean the regionInfo ? any reason why callers of this function would want that ? just the master ip should be sufficient ?

Reply:
I was pointing to the string return which forces us to return a new empty `""` string in case of failures, which can be avoided using a pointer. You can either pass a `*url` / `*string` / `*region` which can avoid this scenario.

Reply:
done